### PR TITLE
Code Release: CARAFE: Content-Aware ReAssembly of FEatures (ICCV 2019)

### PR DESCRIPTION
**Pull request by myownskyW7**

### CODE REVIEW

**Comment by ShihuaiXu** [comment]
Error limit reached.
100 errors detected in the compilation of "/tmp/tmpxft_00006868_00000000-6_carafe_cuda_kernel_benchmark.cpp1.ii".
Compilation terminated.
error: command '/usr/local/cuda/bin/nvcc' failed with exit status 1

File: mmdet/ops/carafe/src/carafe_cuda_kernel.cu [hunk 1]
```diff
@@ -0,0 +1,483 @@
...
+  const int pixel_id = threadIdx.x / THREADS_PER_PIXEL;
+  const int split_id = threadIdx.x % THREADS_PER_PIXEL;
+  // (n, c, ph, pw) is an element in the bottom_data
+  index = index / THREADS_PER_PIXEL;
+  const int pw = index % width;
+  const int ph = (index / width) % height;
+  const int n = index / width / height;
+
+  // const int down_pw = pw / scale_factor;
+  // const int down_ph = ph / scale_factor;
```
Comment (hellock) [inline mmdet/ops/carafe/src/carafe_cuda_kernel.cu]: useless comments

File: mmdet/ops/carafe/src/carafe_naive_cuda_kernel.cu [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+                               const at::Tensor features,
+                               const at::Tensor masks, const int kernel_size,
+                               const int group_size, const int scale_factor,
+                               const int batch_size, const int channels,
+                               const int height, const int width,
+                               at::Tensor bottom_grad, at::Tensor mask_grad) {
+  const int output_size = batch_size * channels * height * width;
+
+  // TODO: use AT_DISPATCH_FLOATING_TYPES_AND_HALF when atomicAdd is resolved
+  AT_DISPATCH_FLOATING_TYPES(
```
Comment (hellock) [inline mmdet/ops/carafe/src/carafe_naive_cuda_kernel.cu]: `AT_DISPATCH_FLOATING_TYPES_AND_HALF` is available here.

**Comment by txytju** [comment]
why not merge to master branch?

File: mmdet/ops/carafe/src/carafe_cuda.cpp [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+int carafe_forward_cuda(at::Tensor features, at::Tensor rfeatures,
+                        at::Tensor masks, at::Tensor rmasks, int kernel_size,
+                        int group_size, int scale_factor, at::Tensor routput,
+                        at::Tensor output) {
+  CHECK_INPUT(features);
+  CHECK_INPUT(rfeatures);
+  CHECK_INPUT(masks);
+  CHECK_INPUT(rmasks);
+  CHECK_INPUT(output);
+  CHECK_INPUT(routput);
```
Comment (hellock) [inline mmdet/ops/carafe/src/carafe_cuda.cpp]: Add a device guard here, like https://github.com/open-mmlab/mmdetection/blob/master/mmdet/ops/roi_align/src/roi_align_cuda.cpp#L36

File: configs/carafe/README.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+Git clone mmdetection
+```shell
+git clone https://github.com/open-mmlab/mmdetection.git
+cd mmdetection
+```
+
+a. Use CARAFE in mmdetection.
+
+Install mmdetection
```
Comment (hellock) [inline configs/carafe/README.md]: It is better not to write the installation command here, since it may be modified in the docs. Just leave a reference to it.
1. Install mmdetection following the official guide.
2. Perform gradient check for CARAFE ops.
```
python ./mmdet/ops/carafe/grad_check.py
```

File: mmdet/models/mask_heads/fcn_mask_head.py [hunk 5]
```diff
@@ -66,14 +70,26 @@ def __init__(self,
...
         elif self.upsample_method == 'deconv':
-            self.upsample = nn.ConvTranspose2d(
-                upsample_in_channels,
-                self.conv_out_channels,
-                self.upsample_ratio,
+            upsampler_cfg_ = dict(
+                in_channels=upsample_in_channels,
+                out_channels=self.conv_out_channels,
+                kernel_size=self.upsample_ratio,
                 stride=self.upsample_ratio)
```
Comment (hellock) [inline mmdet/models/mask_heads/fcn_mask_head.py]: Also add `**self.upsample_cfg` for other options.

File: mmdet/models/utils/upsampler.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        xavier_init(self.upsample_conv, distribution='uniform')
+
+    def forward(self, x):
+        x = self.upsample_conv(x)
+        x = F.pixel_shuffle(x, self.scale_factor)
+        return x
+
+
+upsampler_cfg = {
+    # format: layer_type: (abbreviation, module)
```
Comment (hellock) [inline mmdet/models/utils/upsampler.py]: `layer_type` and `abbreviation` are the same. There is no need to use a tuple for dict values.

File: mmdet/models/utils/upsampler.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    'nearest': ('nearest', nn.Upsample),
+    'bilinear': ('bilinear', nn.Upsample),
+    'deconv': ('deconv', nn.ConvTranspose2d),
+    'pixel_shuffle': ('pixel_shuffle', PixelShufflePack),
+    'carafe': ('carafe', CARAFEPack)
+}
+
+
+def build_upsampler_layer(cfg, postfix=''):
```
Comment (hellock) [inline mmdet/models/utils/upsampler.py]: 1. I suggest using a consistent term "upsample" instead of "upsample" and "upsampler".
2. `postfix` is unnecessary.

File: mmdet/models/utils/upsampler.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn import xavier_init
+
+from mmdet.ops.carafe import CARAFEPack
+
+
+class PixelShufflePack(nn.Module):
```
Comment (hellock) [inline mmdet/models/utils/upsampler.py]: Add docstring here.

File: configs/carafe/faster_rcnn_r50_fpn_carafe_1x.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+        dict(type='TextLoggerHook'),
+        # dict(type='TensorboardLoggerHook')
+    ])
+# yapf:enable
+evaluation = dict(interval=1)
+# runtime settings
+total_epochs = 12
+dist_params = dict(backend='nccl')
+log_level = 'INFO'
+work_dir = './work_dirs/faster_rcnn_r50_fpn_1x'
```
Comment (hellock) [inline configs/carafe/faster_rcnn_r50_fpn_carafe_1x.py]: Use the same name as the config file.

File: mmdet/models/mask_heads/fcn_mask_head.py [hunk 10]
```diff
@@ -20,27 +21,30 @@ def __init__(self,
...
                  num_classes=81,
                  class_agnostic=False,
+                 upsample_cfg=dict(type='deconv', upsample_ratio=2),
                  conv_cfg=None,
                  norm_cfg=None,
                  loss_mask=dict(
                      type='CrossEntropyLoss', use_mask=True, loss_weight=1.0)):
         super(FCNMaskHead, self).__init__()
-        if upsample_method not in [None, 'deconv', 'nearest', 'bilinear']:
+        self.upsample_cfg = upsample_cfg.copy()
```
Comment (hellock) [inline mmdet/models/mask_heads/fcn_mask_head.py]: `upsample_cfg` may not be necessary as an attribute, since we will set `self.upsample_method`, and other fields can be obtained in `self.upsample`.

File: mmdet/models/mask_heads/fcn_mask_head.py [hunk 11]
```diff
@@ -20,27 +21,30 @@ def __init__(self,
...
         self.num_convs = num_convs
         # WARN: roi_feat_size is reserved and not used
         self.roi_feat_size = _pair(roi_feat_size)
         self.in_channels = in_channels
         self.conv_kernel_size = conv_kernel_size
         self.conv_out_channels = conv_out_channels
-        self.upsample_method = upsample_method
-        self.upsample_ratio = upsample_ratio
+        self.upsample_method = self.upsample_cfg.pop('type')
+        self.upsample_ratio = self.upsample_cfg.pop('upsample_ratio')
```
Comment (hellock) [inline mmdet/models/mask_heads/fcn_mask_head.py]: We may consider renaming `upsample_ratio` to `scale_factor` since all upsample operators below have the argument `scale_factor`.

It will be simpler to manipulate the args.
```python
self.upsample_method = upsample_cfg_.get('type')

if self.upsample_method is None:
    self.upsample = None
elif self.upsample_method == 'deconv':
    upsample_cfg_.update(
        in_channels=upsample_in_channels,
        out_channels=self.conv_out_channels,
        kernel_size=self.upsample_ratio,
        stride=self.upsample_ratio)
elif self.upsample_method == 'carafe':
    upsampler_cfg_.update(
        channels=upsample_in_channels,
        scale_factor=self.upsample_ratio)
else:
    xxxxx
```

File: mmdet/models/utils/upsample.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn import xavier_init
+
+from mmdet.ops.carafe import CARAFEPack
+
+
+class PixelShufflePack(nn.Module):
+    """ pixel shuffle upsample layer
```
Comment (hellock) [inline mmdet/models/utils/upsample.py]: `"""Pixel Shuffle upsample layer`

File: mmdet/models/utils/upsample.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+    Args:
+        in_channels (int): number of input channels
```
Comment (hellock) [inline mmdet/models/utils/upsample.py]: The description is a sentence and use upper case for the first character.

File: mmdet/models/utils/upsample.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        out_channels (int): number of output channels
+        scale_factor (int): upsample ratio
+        upsample_kernel (int): kernel size of Conv layer to expand the channels
+    Returns:
```
Comment (hellock) [inline mmdet/models/utils/upsample.py]: Add a blank line before `Returns`.

File: mmdet/ops/carafe/carafe.py [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+                             rgrad_masks, grad_input, grad_masks)
+        return grad_input, grad_masks, None, None, None, None
+
+
+carafe = CARAFEFunction.apply
+
+
+class CARAFE(Module):
+    """ CARAFE: Content-Aware ReAssembly of FEatures
+    Please refer to https://arxiv.org/abs/1905.02188 for more details.
```
Comment (hellock) [inline mmdet/ops/carafe/carafe.py]: Add a blank line between the summary and detailed description.

File: mmdet/ops/carafe/carafe.py [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+
+    Args:
+        channels (int): input feature channels
+        scale_factor (int): upsample ratio
+        up_kernel (int): kernel size of CARAFE op
+        up_group (int): group size of CARAFE op
+        encoder_kernel (int): kernel size of content encoder
+        encoder_dilation (int): dilation of content encoder
+        compressed_channels (int): output channels of channels compressor
+    Returns:
```
Comment (hellock) [inline mmdet/ops/carafe/carafe.py]: Add a blank line before `Returns`.

File: mmdet/ops/carafe/carafe.py [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+        upsampled feature map
+    """
+
+    def __init__(self, kernel_size, group_size, scale_factor):
+        super(CARAFE, self).__init__()
+
+        self.kernel_size = int(kernel_size)
+        self.group_size = int(group_size)
+        self.scale_factor = int(scale_factor)
```
Comment (hellock) [inline mmdet/ops/carafe/carafe.py]: Add a type check instead of simply casting to int.
